File: Java/LinkedList/Easy/ReverseLinkedList.java
package Java.LinkedList.Easy;

public class ReverseLinkedList {

    static class Node {
        int data;
        Node next;
        Node(int data) { this.data = data; }
    }

    static Node insertAtEnd(Node head, int value) {
        Node newNode = new Node(value);
        if (head == null) return newNode;

        Node temp = head;
        while (temp.next != null) temp = temp.next;
        temp.next = newNode;
        return head;
    }

    static Node reverse(Node head) {
        Node prev = null, current = head, next = null;
        while (current != null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        return prev;
    }

    static void display(Node head) {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " -> ");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        Node head = null;
        head = insertAtEnd(head, 10);
        head = insertAtEnd(head, 20);
        head = insertAtEnd(head, 30);

        display(head);
        head = reverse(head);
        display(head);
    }
}

File: Swift/LinkedList/ReverseLinkedList.swift
class Node {
    var data: Int
    var next: Node?
    init(_ data: Int) {
        self.data = data
        self.next = nil
    }
}

var head: Node? = nil

func insert(_ val: Int) {
    let newNode = Node(val)
    if head == nil {
        head = newNode
        return
    }
    var temp = head
    while temp?.next != nil {
        temp = temp?.next
    }
    temp?.next = newNode
}

func reverseList() {
    var prev: Node? = nil
    var curr = head
    while curr != nil {
        let nextTemp = curr?.next
        curr?.next = prev
        prev = curr
        curr = nextTemp
    }
    head = prev
}

func display() {
    var temp = head
    while temp != nil {
        print("\(temp!.data) -> ", terminator: "")
        temp = temp?.next
    }
    print("nil")
}

insert(10)
insert(20)
insert(30)
insert(40)

display()
reverseList()
display()
